server/util.go

```diff
@@ -41,43 +46,62 @@ func NewContext(v *viper.Viper, config *tmcfg.Config, logger log.Logger) *Contex
-// where the logger and config are set based on the execution of parsing or
-// creating a new Tendermint configuration file (config.toml). The provided
-// viper object must be created at the root level and have all necessary flags,
-// defined by Tendermint, bound to it.
-func PersistentPreRunEFn(ctx *Context) func(*cobra.Command, []string) error {
-	return func(cmd *cobra.Command, args []string) error {
-		rootViper := viper.New()
-		rootViper.BindPFlags(cmd.Flags())
-		rootViper.BindPFlags(cmd.PersistentFlags())
-
-		if cmd.Name() == version.Cmd.Name() {
-			return nil
-		}
+// InterceptConfigsPreRunHandler performs a pre-run function for the root daemon
+// application command. It will create a Viper literal and a default server
+// Context. The server Tendermint configuration will either be read and parsed
+// or created and saved to disk, where the server Context is updated to reflect
+// the Tendermint configuration. The Viper literal is used to read and parse
+// the application configuration. Command handlers can fetch the server Context
+// to get the Tendermint configuration or to get access to Viper.
+func InterceptConfigsPreRunHandler(cmd *cobra.Command) error {
+	rootViper := viper.New()
+	rootViper.BindPFlags(cmd.Flags())
+	rootViper.BindPFlags(cmd.PersistentFlags())
+
+	serverCtx := NewDefaultContext()
+	config, err := interceptConfigs(serverCtx, rootViper)
+	if err != nil {
+		return err
+	}
 
-		config, err := interceptConfigs(ctx, rootViper)
-		if err != nil {
-			return err
-		}
+	logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+	logger, err = tmflags.ParseLogLevel(config.LogLevel, logger, tmcfg.DefaultLogLevel())
+	if err != nil {
+		return err
+	}
 
-		logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
-		logger, err = tmflags.ParseLogLevel(config.LogLevel, logger, tmcfg.DefaultLogLevel())
-		if err != nil {
-			return err
-		}
+	if rootViper.GetBool(tmcli.TraceFlag) {
+		logger = log.NewTracingLogger(logger)
+	}
 
-		if rootViper.GetBool(tmcli.TraceFlag) {
-			logger = log.NewTracingLogger(logger)
-		}
+	serverCtx.Config = config
+	serverCtx.Logger = logger.With("module", "main")
+
+	return SetCmdServerContext(cmd, serverCtx)
+}
+
+// GetServerContextFromCmd returns a Context from a command or an empty Context
+// if it has not been set.
+func GetServerContextFromCmd(cmd *cobra.Command) *Context {
+	if v := cmd.Context().Value(ServerContextKey); v != nil {
+		serverCtxPtr := v.(*Context)
+		return serverCtxPtr
+	}
 
-		logger = logger.With("module", "main")
-		ctx.Config = config
-		ctx.Logger = logger
+	return NewDefaultContext()
+}
 
-		return nil
+// SetCmdServerContext sets a command's Context value to the provided argument.
+func SetCmdServerContext(cmd *cobra.Command, serverCtx *Context) error {
+	v := cmd.Context().Value(ServerContextKey)
+	if v == nil {
+		return errors.New("server context not set")
 	}
+
+	serverCtxPtr := v.(*Context)
```

Comment:
this will panic if the type is invalid, should we return an error instead?

Reply:
No, it's guaranteed to be that type.